perf(changelog): cache commit retain checks (258 times faster generation)

aminya opened the pull request:
## Description

This caches the commit retain checks whenever include/exclude paths are specified.

This speeds up changelog generation by `258` times in a big repository I have:
```
Now: 0.080 s
Before:  20.633 s
```

Fixes #767 

It also improves handling including/excluding patterns, considering the first commit, and various other improvements.


## How Has This Been Tested?

I tested this on the repository mentioned in #767

## Screenshots / Logs (if applicable)

Here are the profiling flamegraphs created via #768. The bottleneck is removed after this fix is applied as seen in the above image (4366 samples to only 16 samples). After the fix, other parts of the code finally show up in the profiler.

After - Fully Cached:

![git-cliff - after cached - flamegraph](https://github.com/user-attachments/assets/d29339a5-1c82-4630-bcb8-0b3466d8710a)


After - Cached + new commit:

![git-cliff - after new commit - flamegraph](https://github.com/user-attachments/assets/104866b8-38cf-499d-b904-7409d4d5a938)


Before:

![git-cliff - Before - flamegraph](https://github.com/user-attachments/assets/e7cff5e2-0d69-4b9d-beac-9b6939f23553)




## Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (no code change)
- [ ] Refactor (refactoring production code)
- [x] Other <!--- (provide information) -->

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] My code follows the code style of this project.
- [ ] I have updated the documentation accordingly.
- [x] I have formatted the code with [rustfmt](https://github.com/rust-lang/rustfmt).
- [x] I checked the lints with [clippy](https://github.com/rust-lang/rust-clippy).
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

<!--- Thank you for contributing to git-cliff! ⛰️  -->

alerque commented:
Only a measly 87× faster? Pshaw nothing to see here moving along now…

aminya commented:
> Only a measly 87× faster? Pshaw nothing to see here moving along now…

Challenge accepted. Now 258 times faster. 😄

aminya commented:
I changed the cache implementation to be more useful by caching the changed files of each commit.

This slows down the retain_commit check by ~2 times but allows reusing the cache in case the include/exclude patterns change. This is beneficial for running in monorepos where git-cliff runs multiple times to generate changelog for different subsets of the repo. This is a good trade-off in my opinion.

orhun commented on git-cliff-core/src/repo.rs:
```diff
@@ -76,36 +86,104 @@ impl Repository {
+		let result = self.commit_changed_files_no_cache(commit);
+
+		// Set the result to the cache.
+		match bincode::encode_to_vec(&result, bincode::config::standard()) {
```
Just curious why we need `bincode` here

aminya replied:
Bincode helps with serializing the paths so that they can be stored in the cache. I tried serde_json first, but the whole generation ended up being 2.5 times slower than using bincode. So, I chose bincode to avoid adding too much caching overhead.

orhun reviewed (commented):
🔥

aminya commented on git-cliff-core/src/repo.rs:
```diff
@@ -36,15 +36,25 @@ pub struct Repository {
+			let inner = GitRepository::open(&path)?;
+
+			// hash the path to create a unique cache id
+			let changed_files_cache_path =
+				inner.path().join("git-cliff").join("changed_files_cache");
```
It is saved under `.git/git-cliff/changed_files_cache`. That's the best place for it in my opinion as it is local to the repository. We can also use this path in future if will want to store more things.

aminya commented on git-cliff-core/Cargo.toml:
```diff
@@ -52,6 +52,7 @@ lazy_static.workspace = true
 thiserror = "1.0.61"
 serde = { version = "1.0.203", features = ["derive"] }
 serde_json = "1.0.118"
+bincode = "2.0.0-rc.3"
```
Since the Rc version has been in continuous development, and since it has been in the RC state for quite some time, it works stable. `v1` looks stale (released 3 years ago).
 
There are some breaking changes between the two, and I chose the Rc to avoid needing to upgrade this later once the final tag comes.
https://github.com/bincode-org/bincode/releases

orhun replied:
Makes sense, thanks for the explanation!

aminya commented:
I fixed some bugs in the last three commits to improve the include/exclude behaviour:
- The include/exclude patterns are considered together.
- The glob patterns are normalized so that `./` doesn't make a difference in the result
- Added special case for the first commit where it has no parent

orhun commented on git-cliff-core/src/repo.rs:
```diff
@@ -75,37 +85,152 @@ impl Repository {
+		}
+	}
+
+	fn commit_changed_files(&self, commit: &Commit) -> Vec<PathBuf> {
```
Can you add comments explaining what this function and `_no_cache` variant is doing? It's hard to figure it out while reading it through.

aminya replied:
Added docs in [`7c63698` (#772)](https://github.com/orhun/git-cliff/pull/772/commits/7c6369876cb1ea5ad9ba85d17ba1c98677b020ca)

orhun commented on git-cliff-core/src/repo.rs:
```diff
@@ -75,37 +85,152 @@ impl Repository {
+	}
+
+	/// Calculates whether the commit should be retained or not.
+	fn should_retain_commit(
```
Can you add unit tests for this?

aminya replied:
Added a comprehensive test suit in [`33f56cf` (#772)](https://github.com/orhun/git-cliff/pull/772/commits/33f56cf2d916106168afc5b21ec8685c573304bf)

orhun commented on git-cliff-core/src/repo.rs:
```diff
@@ -75,37 +85,152 @@ impl Repository {
 	}
 
+	/// Normalizes the glob pattern by removing the leading `./`.
+	fn normalize_glob(pattern: Pattern) -> Pattern {
```
I feel like this can be moved into `commits` function (as a inner function maybe) since it is not used anywhere + not being tested.

aminya replied:
It's been used in the tests now as well, so it's better to keep it separate.

orhun reviewed (approved):
LGTM, can you resolve conflicts caused by #709 ? 😇

orhun reviewed (approved):
Awesome, thanks!
